### ISSUE TITLE
"No mailbox with new mail." message clobbered.

### ISSUE THREAD

**Issue by alejandro-colomar** [neomutt/neomutt#4451]
## NeoMutt Version

```
alx@devuan:~$ neomutt -v | head -n1
NeoMutt 20220429
```

But it happens with all versions I've tried so far.

## Expected Behaviour

```
alx@devuan:~$ mutt -Z
No mailbox with new mail.
alx@devuan:~$
```

## Actual Behaviour

```
alx@devuan:~$ neomutt -Z
alx@devuan:~$
```

## Steps to Reproduce

There needs to be no new mail, which in my system, I measure this way:

```
alx@devuan:~$ howmuchmail 
0
alx@devuan:~$ which howmuchmail
/usr/local/bin/howmuchmail
alx@devuan:~$ which howmuchmail | xargs cat
#!/bin/sh
find /home/alx/mail/ -type d \
| grep /new$ \
| xargs -I {} find {} -type f \
| wc -l;
```

## How often does this happen?

- Always

## When did it start to happen?

- Whe I switched from mutt(1) to neomutt(1).

## Extra Info

I had always believed that neomutt(1) was not printing the message.
Today I've realized that neomutt(1) actually prints the message, but it clobbers it with the prompt.
I've realized about this by running neomutt(1) in the console (ctrl+alt+F1).

* Operating System and its version

   Reproducible in both Devuan and Debian stable.

* Were you using multiple copies of NeoMutt at once?

   No

* Were you using 'screen' or 'tmux'?

   No.

* Is your email local (maildir) or remote (IMAP)?

   Local (maildir).  mbsync(1) FTW!  ;)

**Comment by gahr** [comment]
It looks like we're not resetting the terminal correctly

**Comment by gahr** [comment]
I think, in my case, the way why it wasn't refreshing the screen is because I had macro keys waiting (yes, I have a `push AA` line in my config, for reasons...). This causes us not to trigger a `refresh()` that would clean up the screan, here:
https://github.com/neomutt/neomutt/blob/311dd98e98d10f60e86acab0b7ce725a05cad975/gui/curs_lib.c#L84-L89

Can you please double check if you're running into the same use case, i.e., `MacroEvents` is not empty on that line?

**Comment by gahr** [comment]
We can fix it with 

```diff
diff --git a/main.c b/main.c
index 276bf7acf..c5dbd6d5b 100644
--- a/main.c
+++ b/main.c
@@ -1322,6 +1322,7 @@ main
       const CheckStatsFlags csflags = MUTT_MAILBOX_CHECK_IMMEDIATE;
       if (mutt_mailbox_check(NULL, csflags) == 0)
       {
+        OptForceRefresh = true;
         mutt_message(_("No mailbox with new mail"));
         goto main_curses; // TEST37: neomutt -Z (no new mail)
       }
```

but I feel this is adding mess to what's already pretty messy, i.e., our error/screen handling in main.

**Comment by gahr** [comment]
Or perhaps we could stick `OptForceRefresh = true` in `mutt_endwin`... what do you think @flatcap?

**Comment by alejandro-colomar** [comment]
> I think, in my case, the way why it wasn't refreshing the screen is because I had macro keys waiting (yes, I have a `push AA` line in my config, for reasons...). This causes us not to trigger a `refresh()` that would clean up the screan, here:
> 
> https://github.com/neomutt/neomutt/blob/311dd98e98d10f60e86acab0b7ce725a05cad975/gui/curs_lib.c#L84-L89
> 
> Can you please double check if you're running into the same use case, i.e., `MacroEvents` is not empty on that line?

If I comment all `macro ...` lines in the config, it still triggers this bug.  I haven't debugged the program, though.

**Comment by alejandro-colomar** [comment]
> It's more about `push` than `macro`: it's about _calling_ the macro by pushing keys into the input buffer

I don't have any `push` in my config:

```sh
alx@devuan:~/.config/mutt$ grep -rn push
alx@devuan:~/.config/mutt$ 
```

**Comment by gahr** [comment]
It would be great if you could stick a breakpoint in [curs_lib.c](https://github.com/neomutt/neomutt/blob/311dd98e98d10f60e86acab0b7ce725a05cad975/gui/curs_lib.c#L85) and see why `mutt_refresh` doesn't get to the call to `refresh`, if indeed it does not, as I conjecture. Thanks!

**Comment by alejandro-colomar** [comment]
Here's the debug session:

```diff
alx@devuan:~/src/mutt/neomutt/main$ git show
commit 934935242a5592d3c9144a880fef40d537caa971 (HEAD -> main)
Author: Alejandro Colomar (@alejandro-colomar) <alx@kernel.org>
Date:   Mon Dec 9 13:48:55 2024 +0100

    dbg

diff --git a/gui/curs_lib.c b/gui/curs_lib.c
index 6e562ba4f..f38759dbe 100644
--- a/gui/curs_lib.c
+++ b/gui/curs_lib.c
@@ -77,10 +77,13 @@ void mutt_beep(bool force)
  */
 void mutt_refresh(void)
 {
+  fprintf(stderr, "ALX: OptKeepQuiet: %d\n", OptKeepQuiet);
   /* don't refresh when we are waiting for a child. */
   if (OptKeepQuiet)
     return;
 
+  fprintf(stderr, "ALX: !ARRAY_EMPTY(&MacroEvents): %d\n", !ARRAY_EMPTY(&MacroEvents));
+  fprintf(stderr, "ALX: !OptForceRefresh: %d\n", !OptForceRefresh);
   /* don't refresh in the middle of macros unless necessary */
   if (!ARRAY_EMPTY(&MacroEvents) && !OptForceRefresh)
     return;
```
```sh
alx@devuan:~/src/mutt/neomutt/main$ ./neomutt -Z
ALX: OptKeepQuiet: 0
ALX: !ARRAY_EMPTY(&MacroEvents): 0
ALX: !OptForceRefresh: 1
gpg: Note: secret key 6D748A94A1CB00B8 expired at Fri Feb  2 19:17:14 2024 CET
ALX: OptKeepQuiet: 0
ALX: !ARRAY_EMPTY(&MacroEvents): 0
ALX: !OptForceRefresh: 1
```

**Comment by alejandro-colomar** [comment]
If I run this in the console (`/dev/tty1`), I see this, at the bottom of the screen (all abote it is black):

```shell
Option mixmaster is deprecatedALX OptKeepQuiet: 0
                                                 ALX: !ARRAY_EMPTY(&MacroEvents): 0
                                                                                   ALX: !OptForceRefresh: 1
No mailbox with new mailALX: OptKeepQuiet: 0
                                            ALX: !ARRAY_EMPTY(&MacroEvents): 0
                                                                              ALX: !OptForceRefresh: 1
alx@devuan:~/src/mutt/neomutt/main$
```

**Comment by gahr** [comment]
Where does the gpg message come from?

**Comment by alejandro-colomar** [comment]
> Where does the gpg message come from?

```sh
alx@devuan:~/.config/mutt$ grep -rn 'pass '
pass.muttrc:1:set my_migadu_pass	= "`pass show email/migadu/alx/mailbox/alx/pass`"
pass.muttrc:2:set my_tere_pass	= "`pass show email/migadu/alx/mailbox/tere/pass`"
pass.muttrc:3:set my_korg_pass	= "`pass show www/foss/kernel/smtp/pass`"
```

From using the standard Unix password manager.  :)

<https://www.passwordstore.org/>

**Comment by gahr** [comment]
I see.. can you try removing that output?

**Comment by alejandro-colomar** [comment]
> I see.. can you try removing that output?

Sure.  I've hardcoded the passwords in the config now.

Same exact output (except for gpg(1)'s error message):

```sh
$ ./neomutt -Z
ALX: OptKeepQuiet: 0
ALX: !ARRAY_EMPTY(&MacroEvents): 0
ALX: !OptForceRefresh: 1
ALX: OptKeepQuiet: 0
ALX: !ARRAY_EMPTY(&MacroEvents): 0
ALX: !OptForceRefresh: 1
```

**Comment by flatcap** [comment]
I've pushed a fix to [[flatcap/new-mail]](https://github.com/neomutt/neomutt/tree/flatcap/new-mail).

It works, but I don't know if it's the _right_ fix yet

**Comment by flatcap** [comment]
@alejandro-colomar 
You've got a small window, before I make a release.
Does my branch help?

**Comment by alejandro-colomar** [comment]
> @alejandro-colomar You've got a small window, before I make a release. Does my branch help?

I haven't had the time to test it.  I'll do it today.  I'll let you know later today.  :)

**Comment by alejandro-colomar** [comment]
Confirmed: that branch fixes my issue.

```sh
alx@devuan:~/src/mutt/neomutt/main$ ./neomutt -Z
No mailbox with new mail
alx@devuan:~/src/mutt/neomutt/main$ git log -1 --oneline
0a0387b1b (HEAD, neomutt/flatcap/new-mail) fix: no new mail message
```

**Comment by flatcap** [comment]
Fixed by: 3c0a9ec12 fix: no new mail message
